File: views/schiffeVersenken/schiffeVersenken.java
class SchiffeVersenken {
    final String ID;
    final int breite, hoehe;
    final String libPath = "views/schiffeVersenken/schiffeVersenken.js"; // Korrektur: Pfad zum Ordner 'schiffeVersenken'
    LiveView view;

    char[][] spielfeldSpieler = new char[10][10];  // Spielfeld des Spielers (10x10)
    char[][] spielfeldKI = new char[10][10];  // Spielfeld der KI (10x10)
    int[] schiffLaengen = {2, 3, 4, 5}; // Schiffe mit Längen 2, 3, 4 und 5
    int platzierteSchiffeSpieler = 0;  // Zähler für platzierte Schiffe des Spielers
    boolean kiSchiffePlatziert = false;  // Status, ob die KI ihre Schiffe platziert hat

    // Konstruktor
    SchiffeVersenken() {
        this.view = Clerk.view();
        this.breite = 700;  // Standard-Breite des Canvas
        this.hoehe = 800;   // Standard-Höhe des Canvas

        // Lade die JavaScript-Datei, um das SchiffeVersenken-Spiel im Frontend zu initialisieren
        Clerk.load(view, libPath);

        // Erhalte die eindeutige ID für dieses SchiffeVersenken-Objekt
        ID = Clerk.getHashID(this);

        // Canvas für das Spielfeld erstellen und einfügen
        Clerk.write(view, "<canvas id='battleshipCanvas" + ID + "' width='" + this.breite + "' height='" + this.hoehe + "' style='border:1px solid #000;'></canvas>");
        Clerk.script(view, "const schiff" + ID + " = new SchiffeVersenken(document.getElementById('battleshipCanvas" + ID + "'), 'schiff" + ID + "');");

        // Spielfeld initialisieren
        initialisiereSpielfeld();

        // Sobald das Spielfeld initialisiert ist, das Spiel aufrufen
        spieleInitialisieren();
    }

    // Spielfeld initialisieren (mit Wasserzeichen '~')
    void initialisiereSpielfeld() {
        for (int i = 0; i < 10; i++) {
            for (int j = 0; j < 10; j++) {
                spielfeldSpieler[i][j] = '~'; // Alle Felder sind leer
                spielfeldKI[i][j] = '~'; // Alle Felder sind leer
            }
        }
    }

    // Methode, um das Spiel zu initialisieren und das Spielfeld darzustellen
    private void spieleInitialisieren() {
        // Zeige das Spielfeld für den Spieler im Frontend an
        for (int i = 0; i < 10; i++) {
            for (int j = 0; j < 10; j++) {
                // Beispiel: Visualisiere jedes Spielfeld (Optional: Setze ein Wasserzeichen)
                // Hier wird jedes Feld visualisiert, im Beispiel könnte es ein 'Wasser'-Zeichen '~' sein
                schiffePlatzierenSpielerCanvas(i, j, false);  // Beispiel: Vertikale Platzierung als Standard
            }
        }
    }

    // Methode zur Platzierung von Schiffsmarkierungen für den Spieler auf dem Canvas (visualisieren)
    private void schiffePlatzierenSpielerCanvas(int x, int y, boolean horizontal) {
        // Beispielcode für das Zeichnen auf dem Canvas
        Clerk.script(view, "schiff" + ID + ".zeichneSchiff(" + x + ", " + y + ", " + (horizontal ? "true" : "false") + ");");
    }

    // Methode zum Platzieren eines Schiffs auf dem Spielfeld des Spielers
    public boolean schiffPlatzierenSpieler(int zeile, int spalte, boolean horizontal) {
        if (platzierteSchiffeSpieler >= schiffLaengen.length) {
            System.out.println("Alle Schiffe wurden bereits platziert.");
            return false;
        }

        int laenge = schiffLaengen[platzierteSchiffeSpieler];
        if (schiffPlatzieren(spielfeldSpieler, zeile, spalte, laenge, horizontal)) {
            platzierteSchiffeSpieler++;

            // Wenn der Spieler alle Schiffe platziert hat, platziert die KI automatisch ihre Schiffe
            if (platzierteSchiffeSpieler == schiffLaengen.length && !kiSchiffePlatziert) {
                schiffePlatzierenKI();
            }
            return true;
        }
        return false;
    }

    // Hilfsmethode zum Platzieren eines Schiffs auf einem Spielfeld
    private boolean schiffPlatzieren(char[][] spielfeld, int xStart, int yStart, int laenge, boolean horizontal) {
        if (horizontal) {
            if (yStart + laenge > 10) {
                System.out.println("Schiff passt nicht horizontal");
                return false;  // Schiff passt nicht horizontal
            }
            for (int i = 0; i < laenge; i++) {
                if (spielfeld[xStart][yStart + i] != '~') {
                    System.out.println("Überschneidung mit einem anderen Schiff");
                    return false;  // Überschneidung mit einem anderen Schiff
                }
            }
            for (int i = 0; i < laenge; i++) {
                spielfeld[xStart][yStart + i] = 'S';  // 'S' bedeutet ein Teil eines Schiffs
                schiffePlatzierenSpielerCanvas(xStart, yStart + i, true);
            }
        } else {
            if (xStart + laenge > 10) {
                System.out.println("Schiff passt nicht vertikal");
                return false;  // Schiff passt nicht vertikal
            }
            for (int i = 0; i < laenge; i++) {
                if (spielfeld[xStart + i][yStart] != '~') {
                    System.out.println("Überschneidung mit einem anderen Schiff");
                    return false;  // Überschneidung mit einem anderen Schiff
                }
            }
            for (int i = 0; i < laenge; i++) {
                spielfeld[xStart + i][yStart] = 'S';  // 'S' bedeutet ein Teil eines Schiffs
                schiffePlatzierenSpielerCanvas(xStart + i, yStart, false);
            }
        }
        return true;
    }

    // Interne Methode zur automatischen Platzierung der Schiffe für die KI
    private void schiffePlatzierenKI() {
        Random random = new Random();

        for (int i = 0; i < schiffLaengen.length; i++) {
            int laenge = schiffLaengen[i];
            boolean erfolgreich = false;

            // Wiederhole, bis das Schiff erfolgreich platziert wurde
            while (!erfolgreich) {
                int xStart = random.nextInt(10);
                int yStart = random.nextInt(10);
                boolean horizontal = random.nextBoolean();

                erfolgreich = schiffPlatzieren(spielfeldKI, xStart, yStart, laenge, horizontal);
            }
        }
        kiSchiffePlatziert = true;  // Status setzen, dass die KI ihre Schiffe platziert hat
    }
}

File: views/schiffeVersenken/schiffeVersenken.js
class SchiffeVersenken {
    constructor(canvasElement, instanceID) {
        this.canvas = canvasElement;
        this.context = this.canvas.getContext("2d");
        this.instanceID = instanceID;
        this.feldGroesse = 50; // Jede Zelle auf dem Spielfeld ist 50x50 Pixel groß
        this.spielfeldSpieler = []; // Spielfeld des Spielers (10x10)
        this.spielfeldKI = []; // Spielfeld der KI (10x10)
        this.initSpielfeld();
    }

    // Initialisiert das Spielfeld (10x10)
    initSpielfeld() {
        // Initialisiere Spielfeld für Spieler und KI
        for (let i = 0; i < 10; i++) {
            let zeileSpieler = [];
            let zeileKI = [];
            for (let j = 0; j < 10; j++) {
                zeileSpieler.push("~"); // '~' bedeutet Wasser
                zeileKI.push("~"); // '~' bedeutet Wasser
            }
            this.spielfeldSpieler.push(zeileSpieler);
            this.spielfeldKI.push(zeileKI);
        }
        this.zeichneSpielfeld(); // Zeichne das Spielfeld für beide (Spieler und KI)
    }

    // Zeichnet das Spielfeld auf dem Canvas (sowohl für den Spieler als auch für die KI)
    zeichneSpielfeld() {
        this.context.clearRect(0, 0, this.canvas.width, this.canvas.height); // Löscht das Canvas

        // Zeichne das Spielfeld des Spielers
        for (let i = 0; i < 10; i++) {
            for (let j = 0; j < 10; j++) {
                this.zeichneFeld(i, j, 0); // Spielerfeld oben
            }
        }

        // Zeichne das Spielfeld der KI
        for (let i = 0; i < 10; i++) {
            for (let j = 0; j < 10; j++) {
                this.zeichneFeld(i, j, 1); // KI-Feld unten
            }
        }
    }

    // Zeichnet ein einzelnes Feld (Zelle) auf dem Spielfeld
    zeichneFeld(x, y, feldTyp) {
        let status = (feldTyp === 0) ? this.spielfeldSpieler[x][y] : this.spielfeldKI[x][y];

        // Bestimme die Farbe
        if (status === "~") {
            this.context.fillStyle = "#87CEEB"; // Wasserfarbe
        } else if (status === "S") {
            this.context.fillStyle = "#8B4513"; // Schifffarbe
        } else if (status === "X") {
            this.context.fillStyle = "#FF0000"; // Treffer
        } else if (status === "O") {
            this.context.fillStyle = "#A9A9A9"; // Verfehlt
        }

        // Berechne die Position für das Feld (Berücksichtige, dass die KI das untere Spielfeld hat)
        let offsetX = feldTyp === 0 ? 0 : 10; // Spieler ist oben, KI unten
        let offsetY = feldTyp === 0 ? 0 : 10;

        // Zeichne das Rechteck
        this.context.fillRect(y * this.feldGroesse, (x + offsetX) * this.feldGroesse, this.feldGroesse, this.feldGroesse);
        this.context.strokeRect(y * this.feldGroesse, (x + offsetY) * this.feldGroesse, this.feldGroesse, this.feldGroesse); // Rand des Feldes
    }

    // Zeichnet ein Schiff auf dem Spielfeld (in einem bestimmten Feld)
    zeichneSchiff(x, y, horizontal) {
        this.spielfeldSpieler[x][y] = "S"; // 'S' steht für ein Schiff
        this.zeichneSpielfeld(); // Das Spielfeld nach dem Platzieren neu zeichnen

        if (horizontal) {
            // Zeichne horizontal weiter
            for (let i = 1; i < 5; i++) { // Beispiel: 5 Zellen Schiff
                if (y + i < 10) {
                    this.spielfeldSpieler[x][y + i] = "S";
                    this.zeichneSpielfeld();
                }
            }
        } else {
            // Zeichne vertikal weiter
            for (let i = 1; i < 5; i++) { // Beispiel: 5 Zellen Schiff
                if (x + i < 10) {
                    this.spielfeldSpieler[x + i][y] = "S";
                    this.zeichneSpielfeld();
                }
            }
        }
    }

    // Eine Methode zum Setzen eines Schusses auf das Spielfeld (Test)
    schussAbgeben(x, y) {
        if (this.spielfeldKI[x][y] === "S") {
            this.spielfeldKI[x][y] = "X"; // Schiff getroffen
            this.zeichneSpielfeld();
        } else if (this.spielfeldKI[x][y] === "~") {
            this.spielfeldKI[x][y] = "O"; // Verfehlt
            this.zeichneSpielfeld();
        }
    }
}

// Diese Methode ruft das Erstellen des Spielfelds aus dem Java-Backend auf
function createSchiff(id) {
    const canvas = document.getElementById('battleshipCanvas' + id);
    const schiffeVersenkenSpiel = new SchiffeVersenken(canvas, id);
    return schiffeVersenkenSpiel;
}
